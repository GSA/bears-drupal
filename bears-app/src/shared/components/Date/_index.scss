#usagov-bears-app {
  .usa-form-group {
    width: 100%;
  }

  .usa-memorable-date {
    max-width: 320px;
    flex-wrap: wrap;
  }

  .usa-input,
  .usa-select {
    border-radius: 3px;
  }

  /* stylelint-disable selector-class-pattern */
  .usa-form-group--month {
    width: 242px;
  }

  .usa-form-group--day {
    width: 64px;
  }

  .usa-form-group--year {
    width: 95px;
  }
  /* stylelint-enable selector-class-pattern */
}
