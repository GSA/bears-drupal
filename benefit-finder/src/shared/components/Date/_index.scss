@use '../../styles/colors/_index.scss' as color;
@use '../../styles/breakpoints/_index.scss' as *;

#benefit-finder {
  .usa-form-group {
    width: 100%;
  }

  .usa-memorable-date {
    max-width: 320px;
    flex-wrap: wrap;
  }

  .usa-input {
    border-radius: 3px;

    &.usa-input--error {
      outline: 4px solid color.$alert-red !important;
      border: none;
    }
  }

  .usa-form-group--month {
    width: 242px;
  }

  .usa-form-group--day {
    width: 64px;
  }

  .usa-form-group--year {
    width: 95px;
  }

  @media (width >= $desktop) {
    .usa-memorable-date {
      max-width: 100%;
    }
  }
}
