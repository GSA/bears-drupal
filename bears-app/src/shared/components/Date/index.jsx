import PropTypes from 'prop-types'
import parseDate from '../../utils/parseDate'
// import { useHandleClassName } from '../../hooks/useHandleClassName'
import './_index.scss'

/**
 * a parse our date object
 * @component
 * @param {func} onChange - inherited change handler
 * @param {string} required - inherited string value to manage required state
 * @param {object} value - inherited state values
 * @return {Date} returns a tandard format Date ie 1995-12-17T03:24:00
 */
const Date = ({ onChange, value, required, ui }) => {
  const { date, select } = ui
  const { labelDay, labelMonth, labelYear, monthOptions } = date
  const { dateDefaultValue } = select

  // date input values
  // if no value is inherited create a new value
  // or
  // if a value is inherited as an object, return the value
  // or
  // if there is no value, default to an empty string
  // in the case of date we wait until the value from onChange is a 4 digit value
  const monthValue =
    new window.Date(parseDate(value)).getMonth() || (value && value.month) || ''
  const dayValue =
    new window.Date(parseDate(value)).getDate() || (value && value.day) || ''
  const yearValue =
    (value?.year?.length === 4 &&
      new window.Date(parseDate(value)).getFullYear()) ||
    (value && value.year) ||
    ''

  // const defaultClasses =
  //   required === 'TRUE'
  //     ? ['usa-alert', 'usa-alert--error', 'display-none']
  //     : [
  //         'usa-alert',
  //         'usa-alert--info',
  //         `${noBackground ? 'no-background' : ''}`,
  //       ]

  const handleRequired = required === 'TRUE' ? 'required-field' : ''

  const MonthSelect = () => {
    return (
      <div className="usa-form-group usa-form-group--month usa-form-group--select">
        <label className="usa-label" htmlFor="date_of_birth_month">
          {labelMonth}
        </label>
        <select
          className={`usa-select ${handleRequired}`}
          id="date_of_birth_month"
          name="date_of_birth_month"
          aria-describedby="mdHint"
          required="required"
          value={monthValue}
          onChange={onChange}
        >
          <option value="" key="default" disabled>
            {dateDefaultValue}
          </option>
          {monthOptions.map((option, i) => (
            <option value={i} key={`${option.label}-${i}`}>
              {option.label}
            </option>
          ))}
        </select>
      </div>
    )
  }

  const DayInput = () => {
    return (
      <div className="usa-form-group usa-form-group--day">
        <label className="usa-label" htmlFor="date_of_birth_day">
          {labelDay}
        </label>
        <input
          className={`usa-input $${handleRequired}`}
          aria-describedby="mdHint"
          id="date_of_birth_day"
          name="date_of_birth_day"
          maxLength="2"
          pattern="[0-9]*"
          inputMode="numeric"
          value={dayValue}
          onChange={onChange}
        />
      </div>
    )
  }

  const YearInput = () => {
    return (
      <div className="usa-form-group usa-form-group--year">
        <label className="usa-label" htmlFor="date_of_birth_year">
          {labelYear}
        </label>
        <input
          className={`usa-input ${handleRequired}`}
          aria-describedby="mdHint"
          id="date_of_birth_year"
          name="date_of_birth_year"
          minLength="4"
          maxLength="4"
          pattern="[0-9]*"
          inputMode="numeric"
          value={yearValue}
          onChange={onChange}
        />
      </div>
    )
  }

  return (
    <div className="usa-memorable-date">
      <MonthSelect />
      <DayInput />
      <YearInput />
    </div>
  )
}

Date.propTypes = {
  onChange: PropTypes.func,
  required: PropTypes.string,
  value: PropTypes.object,
}

export default Date
