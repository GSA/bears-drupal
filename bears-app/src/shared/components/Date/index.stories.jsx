import Date from './index.jsx'

export default {
  component: Date,
  tags: ['autodocs'],
}

export const Primary = {}
