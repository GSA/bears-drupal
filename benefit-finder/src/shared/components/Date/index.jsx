import PropTypes from 'prop-types'
import { Alert } from '../index'
import './_index.scss'

/**
 * a parse our date object
 * @component
 * @param {func} onChange - inherited change handler
 * @param {string} required - inherited string value to manage required state
 * @param {object} value - inherited state values
 * @return {Date} returns a tandard format Date ie 1995-12-17T03:24:00
 */
const Date = ({ onChange, value, required, ui, id, valid }) => {
  const { date, select } = ui
  const { labelDay, labelMonth, labelYear, monthOptions, alert } = date
  const { dateDefaultValue } = select

  // Note: when we break each input into functional components they trigger unwanted rerenders

  return (
    <div id={`usa-memorable-date-${id}`} className="usa-memorable-date">
      {valid === true && (
        <Alert
          className="date-alert"
          heading={ui.alertBanner.heading}
          description={alert}
          error
        ></Alert>
      )}
      <div className="usa-form-group usa-form-group--month usa-form-group--select">
        <label className="usa-label" htmlFor={`date_of_birth_month-${id}`}>
          {labelMonth}
        </label>
        <div id={`month-description-${id}`} className="a11y-sr-only">
          Select a month from the list
        </div>
        <select
          className={`usa-select ${
            required === 'TRUE' ? 'required-field' : ''
          }`}
          // ref={monthFieldRef}
          id={`date_of_birth_month-${id}`}
          name={`date_of_birth_month-${id}`}
          aria-describedby={`month-description-${id}`}
          required="required"
          value={(value && value.month) || ''}
          onChange={onChange}
        >
          <option value="" key="default" disabled>
            {dateDefaultValue}
          </option>
          {monthOptions.map((option, i) => (
            <option value={i} key={`${option.label}-${i}`}>
              {option.label}
            </option>
          ))}
        </select>
      </div>
      <div className="usa-form-group usa-form-group--day">
        <label className="usa-label" htmlFor={`date_of_birth_day-${id}`}>
          {labelDay}
        </label>
        <div id={`day-description-${id}`} className="a11y-sr-only">
          Enter numerals for day
        </div>
        <input
          className={`usa-input ${required === 'TRUE' ? 'required-field' : ''}`}
          aria-describedby={`day-description-${id}`}
          id={`date_of_birth_day-${id}`}
          name={`date_of_birth_day-${id}`}
          inputMode="numeric"
          value={(value && value.day) || ''}
          onChange={onChange}
          // ref={dayFieldRef}
        />
      </div>
      <div className="usa-form-group usa-form-group--year">
        <label className="usa-label" htmlFor={`date_of_birth_year-${id}`}>
          {labelYear}
        </label>
        <div id={`year-description-${id}`} className="a11y-sr-only">
          Enter numerals for year
        </div>
        <input
          className={`usa-input ${required === 'TRUE' ? 'required-field' : ''}`}
          aria-describedby={`year-description-${id}`}
          id={`date_of_birth_year-${id}`}
          name={`date_of_birth_year-${id}`}
          inputMode="numeric"
          value={(value && value.year) || ''}
          onChange={onChange}
          // ref={yearFieldRef}
        />
      </div>
    </div>
  )
}

Date.propTypes = {
  onChange: PropTypes.func,
  required: PropTypes.string,
  value: PropTypes.object,
}

export default Date
