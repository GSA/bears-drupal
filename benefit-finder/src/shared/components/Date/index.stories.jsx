import Date from './index.jsx'
import * as en from '../../locales/en/en.json'

export default {
  component: Date,
  tags: ['autodocs'],
  args: {
    ui: { ...en },
    value: { month: '2', day: '22', year: '1977' },
  },
}

export const Primary = {}
