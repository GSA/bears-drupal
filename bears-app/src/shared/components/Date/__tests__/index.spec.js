import renderer from 'react-test-renderer'
import Date from '../index.jsx'

describe('Date', () => {
  it('renders a match to the previous snapshot', () => {
    const component = renderer.create(<Date />)
    expect(component.toJSON()).toMatchSnapshot()
  })
})
