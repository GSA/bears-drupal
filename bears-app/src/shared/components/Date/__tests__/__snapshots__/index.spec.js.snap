// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date renders a match to the previous snapshot 1`] = `
<div
  className="usa-memorable-date"
>
  <div
    className="usa-form-group usa-form-group--month usa-form-group--select"
  >
    <label
      className="usa-label"
      htmlFor="date_of_birth_month"
    >
      Month
    </label>
    <select
      aria-describedby="mdHint"
      className="usa-select "
      id="date_of_birth_month"
      name="date_of_birth_month"
      required="required"
      value=""
    >
      <option
        disabled={true}
        value=""
      >
        -Select-
      </option>
      <option
        value={0}
      >
        01 - January
      </option>
      <option
        value={1}
      >
        02 - February
      </option>
      <option
        value={2}
      >
        03 - March
      </option>
      <option
        value={3}
      >
        04 - April
      </option>
      <option
        value={4}
      >
        05 - May
      </option>
      <option
        value={5}
      >
        06 - June
      </option>
      <option
        value={6}
      >
        07 - July
      </option>
      <option
        value={7}
      >
        08 - August
      </option>
      <option
        value={8}
      >
        09 - September
      </option>
      <option
        value={9}
      >
        10 - October
      </option>
      <option
        value={10}
      >
        11 - November
      </option>
      <option
        value={11}
      >
        12 - December
      </option>
    </select>
  </div>
  <div
    className="usa-form-group usa-form-group--day"
  >
    <label
      className="usa-label"
      htmlFor="date_of_birth_day"
    >
      Day
    </label>
    <input
      aria-describedby="mdHint"
      className="usa-input $"
      id="date_of_birth_day"
      inputMode="numeric"
      maxLength="2"
      name="date_of_birth_day"
      pattern="[0-9]*"
      value=""
    />
  </div>
  <div
    className="usa-form-group usa-form-group--year"
  >
    <label
      className="usa-label"
      htmlFor="date_of_birth_year"
    >
      Year
    </label>
    <input
      aria-describedby="mdHint"
      className="usa-input "
      id="date_of_birth_year"
      inputMode="numeric"
      maxLength="4"
      minLength="4"
      name="date_of_birth_year"
      pattern="[0-9]*"
      value=""
    />
  </div>
</div>
`;
